Index: ext_autoload.php
===================================================================
--- ext_autoload.php	(revision 1068)
+++ ext_autoload.php	(working copy)
@@ -4,6 +4,7 @@
 
 $extensionClassesPath = t3lib_extMgm::extPath('extbase') . 'Classes/';
 return array(
+	'tx_extbase_backenddispatcher' => $extensionClassesPath . 'BackendDispatcher.php',
 	'tx_extbase_dispatcher' => $extensionClassesPath . 'Dispatcher.php',
 	'tx_extbase_exception' => $extensionClassesPath . 'Exception.php',
 	'tx_extbase_configuration_exception' => $extensionClassesPath . 'Configuration/Exception.php',
@@ -240,4 +241,4 @@
 	'tx_extbase_validation_validator_validatorinterface' => $extensionClassesPath . 'Validation/Validator/ValidatorInterface.php',
 	'tx_extbase_view_emptyview' => $extensionClassesPath . 'View/EmptyView.php',
 );
-?>
\ No newline at end of file
+?>
